File: src/competitiveProgrammingTest/cpp/competitive/programming/physics/DiskTest.cpp
#include "gtest/gtest.h"

#include "competitive/programming/physics/Disk.hpp"

#include <cmath>

using competitive::programming::physics::Disk;

TEST(Disk, Move)
{
	Disk disk(2, 1, 2, 3, 5);
	ASSERT_EQ(Disk(4, 4, 2, 3, 5), disk.move());
	ASSERT_EQ(Disk(2, 1, 2, 3, 5), disk);
	disk.moveInPlace();
	ASSERT_EQ(Disk(4, 4, 2, 3, 5), disk);
}

TEST(Disk, Accelerate)
{
	Disk disk(2, 1, 2, 3, 5);
	ASSERT_EQ(Disk(2, 1, 3, 5, 5), disk.accelerate(Disk::Vector(1,2)));
	ASSERT_EQ(Disk(2, 1, 4, 6, 5), disk.accelerate(2));
	ASSERT_EQ(Disk(2, 1, 2, 3, 5), disk);
	disk.accelerateInPlace(Disk::Vector(2,1));
	ASSERT_EQ(Disk(2, 1, 4, 4, 5), disk);
	disk.accelerateInPlace(0.5);
	ASSERT_EQ(Disk(2, 1, 2, 2, 5), disk);
}

TEST(Disk, CollisionDetection)
{
	Disk disk(2, 1, 2, 3, 5);
	Disk oppositeMoves(-2, -5,-2, -3, 1);
	ASSERT_FALSE(disk.willCollide(oppositeMoves));
	Disk frontCollision(6, 7,-2, -3, 1);
	ASSERT_TRUE(disk.willCollide(frontCollision));
	Disk alreadyCollide(2, 1,-2, -3, 1);
	ASSERT_TRUE(disk.willCollide(alreadyCollide));
	Disk noRelativeMovement(10, 10,2, 3, 2);
	ASSERT_FALSE(disk.willCollide(noRelativeMovement));
	Disk goingRight(0, 0,10, 0, 5);
	Disk radiusTouch(2, 6,0, 0, 1);
	ASSERT_TRUE(goingRight.willCollide(radiusTouch));
	Disk radiusNotTouch(2, 6,0, 0, 0.5);
	ASSERT_FALSE(goingRight.willCollide(radiusNotTouch));
}

TEST(Disk, CollisionTime)
{
	Disk disk(2, 1, 2, 3, 5);

	Disk oppositeMoves(-2, -5,-2, -3, 1);
	bool nan = std::isnan(disk.collisionTime(oppositeMoves));
	ASSERT_TRUE(nan);
	Disk goingRight(0, 0,10, 0, 5);
	Disk radiusTouch(2, 6,0, 0, 1);
	ASSERT_EQ(0.2, goingRight.collisionTime(radiusTouch));
	ASSERT_EQ(2, goingRight.collisionTime(Disk(26, 0,0, 0, 1)));
	ASSERT_EQ(0, goingRight.collisionTime(goingRight));
	ASSERT_EQ(0, goingRight.collisionTime(Disk(1, 1,-10, 0, 5)));
}

File: src/test/java/competitive/programming/physics/DiskTest.java
package competitive.programming.physics;

import static org.junit.Assert.*;

import org.junit.Test;

import competitive.programming.geometry.Vector;
import competitive.programming.physics.Disk;

public class DiskTest {
	private Disk disk = new Disk(new Vector(2, 1), new Vector(2, 3), 5);

	@Test
	public void testMove() {
		assertEquals(new Disk(new Vector(4, 4), new Vector(2, 3), 5), disk.move());
		assertEquals(new Disk(new Vector(2, 1), new Vector(2, 3), 5), disk);
	}

	@Test
	public void testAccelerate() {
		assertEquals(new Disk(new Vector(2, 1), new Vector(3, 5), 5), disk.accelerate(new Vector(1, 2)));
		assertEquals(new Disk(new Vector(2, 1), new Vector(4, 6), 5), disk.accelerate(2));
		assertEquals(new Disk(new Vector(2, 1), new Vector(2, 3), 5), disk);
	}
	
	@Test
	public void testCollisionDetection(){
		Disk oppositeMoves = new Disk(new Vector(-2, -5), new Vector(-2, -3), 1);
		assertFalse(disk.willCollide(oppositeMoves));
		Disk frontCollision = new Disk(new Vector(6, 7), new Vector(-2, -3), 1);
		assertTrue(disk.willCollide(frontCollision));
		Disk alreadyCollide = new Disk(new Vector(2, 1), new Vector(-2, -3), 1);
		assertTrue(disk.willCollide(alreadyCollide));
		Disk noRelativeMovement = new Disk(new Vector(10, 10), new Vector(2, 3), 2);
		assertFalse(disk.willCollide(noRelativeMovement));
		Disk goingRight = new Disk(new Vector(0,0), new Vector(10, 0), 5);
		Disk radiusTouch = new Disk(new Vector(2, 6), new Vector(0, 0), 1);
		assertTrue(goingRight.willCollide(radiusTouch));
		Disk radiusNotTouch = new Disk(new Vector(2, 6), new Vector(0, 0), 0.5);
		assertFalse(goingRight.willCollide(radiusNotTouch));
	}
	
	@Test
	public void collisionTime(){
		Disk oppositeMoves = new Disk(new Vector(-2, -5), new Vector(-2, -3), 1);
		assertEquals(Double.MAX_VALUE, disk.collisionTime(oppositeMoves), 0.01);
		Disk goingRight = new Disk(new Vector(0,0), new Vector(10, 0), 5);
		Disk radiusTouch = new Disk(new Vector(2, 6), new Vector(0, 0), 1);
		assertEquals(0.2, goingRight.collisionTime(radiusTouch), 0.01);
		assertEquals(2, goingRight.collisionTime(new Disk(new Vector(26,0), new Vector(0, 0), 1)), 0.01);
		assertEquals(0, goingRight.collisionTime(goingRight), 0.01);
		assertEquals(0, goingRight.collisionTime(new Disk(new Vector(1,1), new Vector(-10, 0), 5)), 0.01);
	}

}
